lxd/device/device.go

```diff
@@ -0,0 +1,115 @@
+	// It calls the supplied volatile save function if state needs persisting.
+	Stop(VolatileConfig, func(map[string]string) error) error
+
+	// Update performs live host-side modifications for a started device based on the difference
+	// between the current config and previous config supplied as an argument.
+	// VolatileConfig from previous instantiations of this device is required.
+	// It calls the supplied volatile save function if state needs persisting.
+	Update(VolatileConfig, Config, func(map[string]string) error) error
+
+	// UpdateFields returns a list of fields that can be updated whilst the device is started.
+	UpdateFields() []string
+
+	// CanHotPlug returns true if device can be started/stopped whilst instance is running.
+	CanHotPlug() bool
+}
+
+// device represents a sealed interface that implements Device, but also contains some internal
+// setup functions for a Device that should only be called by device.New() to avoid exposing devices
+// that are not in a known configured state. This is separate from the Device interface so that
+// Devices created outside of the device package can be used by LXD, but ensures that any devices
+// created by the device package will only be accessible after being configured properly by New().
+type device interface {
+	Device
+
+	// init stores the instanceIdentifier, daemon State and Config into device and performs any setup.
+	init(instanceIdentifier, *state.State, Config)
+
+	// validate checks Config stored by init() is valid for the instance type.
+	validate() error
+}
+
+// deviceCommon represents the common struct for all devices.
+type deviceCommon struct {
+	instance instanceIdentifier
+	config   map[string]string
+	state    *state.State
+}
+
+// init stores the instanceIdentifier, daemon state and Config into device. This is implemented as
+// part of deviceCommon so that the majority of devices don't need to implement it and can just
+// embed deviceCommon.
+func (d *deviceCommon) init(instance instanceIdentifier, state *state.State, conf Config) {
+	d.instance = instance
+	d.config = conf
+	d.state = state
+}
+
+// Update returns an error as most devices do not support live updates.
+func (d *deviceCommon) Update(v VolatileConfig, oldConfig Config, save func(map[string]string) error) error {
+	return fmt.Errorf("Device does not support updates whilst started")
+}
+
+// UpdateFields returns an empty list of update fields as most devices do not support live updates.
+func (d *deviceCommon) UpdateFields() []string {
+	return []string{}
+}
+
+// CanHotPlug returns true as majority of devices can be started/stopped when instance is running.
+func (d *deviceCommon) CanHotPlug() bool {
+	return true
+}
+
+// New instantiates a new device struct, validates the supplied config and sets it into the device.
+func New(instance instanceIdentifier, state *state.State, devConfig Config) (Device, error) {
+	var d device
+	if devConfig["type"] == "nic" {
```

Comment:
I think we can improve this a bit by having a global pre-set map at the top of the file which maps from the type name to the handler and then within the handler for the sub-types (nictype).

As much as possible, I'd like adding a new device type being as simple as dropping a new file in place and adding one line to a list of device types.

Reply:
@stgraber I've added 2 maps of functions to provide this approach in a way that avoids using reflect, can you take a look and see if what is what you had in mind.